src/npyiter.rs

```diff
@@ -0,0 +1,478 @@
+        }
+    }
+
+    pub fn set(mut self, flag: NpyIterFlag) -> Self {
+        self.flags |= flag.to_c_enum();
+        self
+    }
+
+    pub fn unset(mut self, flag: NpyIterFlag) -> Self {
+        self.flags &= !flag.to_c_enum();
+        self
+    }
+}
+
+impl<'py, T: Element, S: MultiIterMode> NpyMultiIterBuilder<'py, T, S> {
+    pub fn add_readonly_array<D: ndarray::Dimension>(
+        mut self,
+        array: &'py PyArray<T, D>,
+    ) -> NpyMultiIterBuilder<'py, T, RO<S>> {
+        self.arrays.push(array.to_dyn());
+        self.opflags.push(NPY_ITER_READONLY);
+
+        NpyMultiIterBuilder {
+            flags: self.flags,
+            opflags: self.opflags,
+            arrays: self.arrays,
+            structure: PhantomData,
+        }
+    }
+
+    pub fn add_readwrite_array<D: ndarray::Dimension>(
+        mut self,
+        array: &'py PyArray<T, D>,
+    ) -> NpyMultiIterBuilder<'py, T, RW<S>> {
+        self.arrays.push(array.to_dyn());
+        self.opflags.push(NPY_ITER_READWRITE);
+
+        NpyMultiIterBuilder {
+            flags: self.flags,
+            opflags: self.opflags,
+            arrays: self.arrays,
+            structure: PhantomData,
+        }
+    }
+}
+
+impl<'py, T: Element, S: MultiIterModeHasManyArrays> NpyMultiIterBuilder<'py, T, S> {
+    pub fn build(mut self) -> PyResult<NpyMultiIterArray<'py, T, S>> {
+        assert!(self.arrays.len() == self.opflags.len());
+        assert!(self.arrays.len() <= i32::MAX as usize);
+        assert!(2 <= self.arrays.len());
+
+        let iter_ptr = unsafe {
+            PY_ARRAY_API.NpyIter_MultiNew(
+                self.arrays.len() as i32,
+                self.arrays
+                    .iter_mut()
+                    .map(|x| x.as_array_ptr())
+                    .collect::<Vec<_>>()
+                    .as_mut_ptr(),
+                self.flags,
+                NPY_ORDER::NPY_ANYORDER,
+                NPY_CASTING::NPY_SAFE_CASTING,
+                self.opflags.as_mut_ptr(),
+                ptr::null_mut(),
+            )
+        };
+        let py = self.arrays[0].py();
+        NpyMultiIterArray::new(iter_ptr, py).ok_or_else(|| PyErr::fetch(py))
+    }
+}
+
+pub struct NpyMultiIterArray<'py, T, S: MultiIterModeHasManyArrays> {
+    iterator: ptr::NonNull<objects::NpyIter>,
+    iternext: unsafe extern "C" fn(*mut objects::NpyIter) -> c_int,
+    empty: bool,
+    iter_size: npy_intp,
+    dataptr: *mut *mut c_char,
+
+    return_type: PhantomData<T>,
+    structure: PhantomData<S>,
+    _py: Python<'py>,
+}
+
+impl<'py, T, S: MultiIterModeHasManyArrays> NpyMultiIterArray<'py, T, S> {
+    fn new(iterator: *mut objects::NpyIter, py: Python<'py>) -> Option<Self> {
+        let mut iterator = ptr::NonNull::new(iterator)?;
+
+        // TODO replace the null second arg with something correct.
+        let iternext =
+            unsafe { PY_ARRAY_API.NpyIter_GetIterNext(iterator.as_mut(), ptr::null_mut())? };
```

Comment:
Reading [the document](https://numpy.org/doc/stable/reference/c-api/iterator.html?highlight=getiternext#c.NpyIter_GetIterNext), it looks that Python exception is set when the `errmsg` is NULL.
So I think null is OK here though we have to call `PyErr::fetch` if `GetIterNext` returns `None`.

Reply:
Hmm, so I ended up changing the signature of the new method to be a PyResult<...> because the other three points where the function could fail there would be no exception set.